Handwritten parser for parsing requirements

hrnciar opened the pull request:
Hello,

I wrote a parser for packaging that should allow us to remove the dependency on pyparsing. 
I am opening this as a draft as this is my first PR into packaging and I would love to hear feedback from you on what to improve in my PR. 

Thank you.

Fixes: #468

brettcannon commented:
Two quick things (I haven't really reviewed any of the code).

One, looks like there are some CI failures.

Two, do we want `packaging.parser` and `packaging.tokenizer` to be public? They strike me as implementation details, so we would probably prefer to keep them as private (i.e. leading underscore in the module names).

pradyunsg commented:
+1 to both the things that Brett said.

pradyunsg commented:
All that said, this is super cool and I'm very excited to see this progressing! Thanks for picking this up @hrnciar! ^>^

pradyunsg commented on packaging/requirements.py:
```diff
@@ -98,11 +34,12 @@ class Requirement:
     # TODO: Can we normalize the name and extra name?
 
     def __init__(self, requirement_string: str) -> None:
+        Requirement = namedtuple('Requirement', ['name', 'url', 'extras', 'specifier', 'marker'])
```
Yea. It would also make sense to make this a private module-level global, so... `_RequirementTuple` works too. :)

uranusjr commented on packaging/_parser.py:
```diff
@@ -0,0 +1,155 @@
+    NAMED_REQUIREMENT: NAME EXTRAS* URL_SPEC (SEMICOLON + MARKER)*
+    """
+    tokens = Tokenizer(requirement)
+    name = tokens.read("IDENTIFIER").text.strip("'\"")
```
What does the `strip("'\"")` part do? I don't think the NAME part can contain quotes. Should this be `strip()`?

hrnciar replied:
You are right, the `strip()` is unnecessary and can be removed entirely.

hroncok commented on packaging/_parser.py:
```diff
@@ -0,0 +1,155 @@
+from ._tokenizer import Tokenizer
+
+
+def parse_named_requirement(requirement: str) -> Tuple[Any, str, Any, str, str]:
```
Shouldn't this actually be:

```suggestion
def parse_named_requirement(requirement: str) -> Tuple[str, str, str, str, str]:
```

? Similarily in all the other places where the type is set to Any.

hrnciar replied:
In this case, it should be `Tuple[str, str, List[str], str, str]`. I will take a look at the rest of them.

pradyunsg commented on packaging/_parser.py:
```diff
@@ -0,0 +1,155 @@
+from typing import Any, List, Tuple
```
I'd appreciate it if we can get rid of as many of the `Any`s here as feasible. Not a blocker, obviously.

hroncok replied:
I know that @hrnciar is working on redoing the typing annotations not to use Anys. Most of the typing annotation comments from this review will be resolved by his WIP changes.

pradyunsg replied:
And I imagine our mypy lint job is complaining about them as well. :)

pradyunsg commented on packaging/_parser.py:
```diff
@@ -0,0 +1,155 @@
+
+
+def parse_variable(tokens: Tokenizer) -> Any:
+    from .markers import Variable
```
We should move the AST classes to over here, allowing them to be used here without needing a deferred import to avoid import cycles.

pradyunsg commented on packaging/_tokenizer.py:
```diff
@@ -0,0 +1,153 @@
+    (advance to the next token).
+    """
+
+    next_token: Any
```
Isn't this guaranteed to be a token?

pradyunsg commented on packaging/_tokenizer.py:
```diff
@@ -0,0 +1,153 @@
+    next_token: Any
+
+    def __init__(
+        self, source: Sized, rules: Dict[Optional[str], object] = DEFAULT_RULES
```
It's likely reasonable to require source to be a string.

pradyunsg commented on packaging/_tokenizer.py:
```diff
@@ -0,0 +1,153 @@
+        self.generator = self._tokenize()
+        self.position = 0
+
+    def peek(self, *match_args: None, **match_kwargs: None) -> Any:
```
These none annotations are weird. If it's not supposed to be passed, just don't have varargs/kwargs?

pradyunsg replied:
This basically applies to the whole file.

pradyunsg commented on packaging/_tokenizer.py:
```diff
@@ -0,0 +1,153 @@
+        if self.match(*match_args, **match_kwargs):
+            return self.read()
+
+    def raise_syntax_error(self, message: str = "Invalid marker") -> Any:
```
The message should likely not have a default value, and instead be a required kwarg.

pradyunsg commented on packaging/_tokenizer.py:
```diff
@@ -0,0 +1,153 @@
+
+
+class Token:
+    def __init__(self, name: str, text: str, position: int):
```
This is missing a -> None

pradyunsg commented on packaging/_parser.py:
```diff
@@ -0,0 +1,155 @@
+from ._tokenizer import Tokenizer
+
+
+def parse_named_requirement(requirement: str) -> Tuple[str, str, List[str], str, str]:
```
It would be nice to return a namedtuple from here, so that this definition is self-contained.

pradyunsg commented:
Pulling this out of draft status, given there is a review request.

pradyunsg reviewed (commented):
This is pretty much a desk review -- I haven't cloned and poked at this yet; I'll do that once the linters are happy with this PR.

I don't think I have anything to say about the parsing logic other than that it looks correct, and that CI being happy is a good sign. :)

I think we can/should bundle this change with the legacy version removal, which... I'll pick back up once this lands.

pradyunsg commented:
A gentle ping.

encukou commented on packaging/_parser.py:
```diff
@@ -0,0 +1,200 @@
+        return str(self)
+
+
+MarkerType = Tuple[Union[Variable, Value], Op, Union[Variable, Value]]
+# MarkerAtom = Union[MarkerType, List["MarkerAtom"]]
```
These names are a bit confusing. Perhaps use `MarkerAtom`, `MarkerItem`, `MarkerVar` and `MarkerList`, corresponding with their parse rules?

pradyunsg commented:
(FYI: The CI isn't running due to merge conflicts with `main`)

hrnciar commented on packaging/_parser.py:
```diff
@@ -0,0 +1,200 @@
+
+
+class Node:
+    def __init__(self, value: Any) -> None:
```
Yes, you are correct. I replaced it with `str`. Thank you.

hrnciar commented:
> (FYI: The CI isn't running due to merge conflicts with `main`)

Thanks, I rebased PR on top of the main branch.

encukou commented on packaging/_parser.py:
```diff
@@ -0,0 +1,201 @@
+    if tokens.try_read("LBRACKET"):
+        while tokens.match("IDENTIFIER"):
+            extras.append(tokens.read("IDENTIFIER").text)
+            tokens.try_read("COLON")
```
This looks like it'll will allow the comma to be skipped. Please test `spam[eggs ham]`: it should not be valid.

encukou commented on packaging/_parser.py:
```diff
@@ -0,0 +1,201 @@
+            parsed_specifiers += tokens.read("VERSION").text
+        else:
+            tokens.raise_syntax_error(message="Missing version")
+        if tokens.match("COLON"):
```
Similar situation here.

encukou commented on packaging/_parser.py:
```diff
@@ -0,0 +1,201 @@
+
+
+def parse_quoted_marker(tokens: Tokenizer) -> MarkerList:
+    tokens.try_read("SEMICOLON")
```
This function should have a docstring with the grammar.

encukou commented on packaging/_parser.py:
```diff
@@ -0,0 +1,201 @@
+
+def parse_marker_var(tokens: Tokenizer) -> MarkerVar:
+    """
+    MARKER_VAR: VARIABLE MARKER_VALUE
```
```suggestion
    MARKER_VAR: VARIABLE | MARKER_VALUE
```

encukou commented on packaging/_parser.py:
```diff
@@ -0,0 +1,201 @@
+def parse_python_str(tokens: Tokenizer) -> Value:
+
+    if tokens.match("QUOTED_STRING"):
+        python_str = tokens.read().text.strip("'\"")
```
```suggestion
        python_str = tokens.read().text[1:-1]
```

uranusjr replied:
Would it make sense to use `ast.literal_eval` to handle escape sequences? (I think they are technically allowed...? PEP 345 simply says the value can be “a free string”.)

encukou replied:
[PEP 508](https://www.python.org/dev/peps/pep-0508/#grammar) specifies a *very* restricted charset for this, actually. But the old parser uses pyparsing's QuotedString, so `ast.literal_eval` would match the old behavior better.
(That is what we want -- where possible, any behavior changes to better match the spec should not be part of *this* change.)

encukou commented on packaging/_tokenizer.py:
```diff
@@ -0,0 +1,154 @@
+        self.text = text
+        self.position = position
+
+    def matches(self, name: str = "", text: str = "") -> bool:
```
The `text` argument is now unused, can you remove it?

 If yes, `match_args`/`match_kwargs` below can be replaced by `name`.

encukou commented on packaging/_tokenizer.py:
```diff
@@ -0,0 +1,154 @@
+    "LBRACKET": r"\[",
+    "RBRACKET": r"\]",
+    "SEMICOLON": r";",
+    "COLON": r",",
```
```suggestion
    "COMMA": r",",
```

encukou commented on packaging/_tokenizer.py:
```diff
@@ -0,0 +1,154 @@
+                + tuple(f"{k}={v!r}" for k, v in match_kwargs.items())
+                if v
+            )
+            raise self.raise_syntax_error(message=f"Expected {exp}")
```
It turns out that the autogenerated  error messages aren't really used. It might be better if `expect` and `read` took an `error_message` argument.

encukou commented on tests/test_requirements.py:
```diff
@@ -89,21 +89,27 @@ def test_empty_extras(self):
     def test_invalid_url(self):
         with pytest.raises(InvalidRequirement) as e:
             Requirement("name @ gopher:/foo/com")
-        assert "Invalid URL: " in str(e.value)
```
Can the `value` attribute be added back? (Is it really missing?)

encukou commented on tests/test_requirements.py:
```diff
@@ -59,20 +59,20 @@ def test_name_with_version(self):
         req = Requirement("name>=3")
         self._assert_requirement(req, "name", specifier=">=3")
 
-    def test_with_legacy_version(self):
-        req = Requirement("name==1.0.org1")
-        self._assert_requirement(req, "name", specifier="==1.0.org1")
-
-    def test_with_legacy_version_and_marker(self):
-        req = Requirement("name>=1.x.y;python_version=='2.6'")
-        self._assert_requirement(
-            req, "name", specifier=">=1.x.y", marker='python_version == "2.6"'
-        )
```
These specifications should still be tested. Do they cause the right error?

encukou commented on tests/test_requirements.py:
```diff
@@ -89,21 +89,27 @@ def test_empty_extras(self):
+        assert "Closing square bracket is missing" in str(e)
+
     def test_url(self):
-        url_section = "@ http://example.com"
```
The old specifications should still be tested. Do they cause the right error?

hrnciar replied:
I've added a `test_missing_name` to cover this. Yes, it causes the right error.

pradyunsg commented on packaging/requirements.py:
```diff
@@ -31,12 +32,13 @@ class Requirement:
     # TODO: Can we normalize the name and extra name?
 
     def __init__(self, requirement_string: str) -> None:
+        _RequirementTuple = namedtuple(
+            "_RequirementTuple", ["name", "url", "extras", "specifier", "marker"]
+        )
```
This should be computed outside of `__init__` to avoid paying the cost of creating this class each time an instance is created.

uranusjr commented on packaging/_parser.py:
```diff
@@ -0,0 +1,204 @@
+        self.value = value
+
+    def __str__(self) -> str:
+        return str(self.value)
```
```suggestion
        return self.value
```

Since `self.value` is always str

pradyunsg commented on packaging/_parser.py:
```diff
@@ -0,0 +1,204 @@
+
+
+def parse_marker_item(
+    tokens: Tokenizer,
```
not: drop the comma.

pradyunsg commented on packaging/_parser.py:
```diff
@@ -0,0 +1,204 @@
+
+
+def parse_variable(tokens: Tokenizer) -> Variable:
+
```
nit: drop new line at start of function.

pradyunsg replied:
Same for the other instances of this.

pradyunsg reviewed (approved):
LGTM, with only nit picks now.

There’s a few follow ups, like improving the error message on a semicolon that’s missing white space and on legacy versions; but we should do those in a follow up.

A blocker for releasing this is going to be packaging dropping the legacy versions and legacy specifiers though, so… that needs to be done as well. I’ll pick up #407 sometime in the coming week or next year.

hrnciar commented:
I've pushed a change to address the nit picks. Thank you all for your help with this PR.

ssbarnea commented:
@hrnciar Any chance to revive/rebase this? I think that this patch is becoming more important as py311 is getting closer. Thanks!

hroncok commented:
We were told this is blocked on https://github.com/pypa/packaging/pull/407

pradyunsg commented on packaging/specifiers.py:
```diff
@@ -341,10 +343,10 @@ class Specifier(_IndividualSpecifier):
                 (?:
+                    \.\*  # Wild card syntax of .*
+                    |
                     (?:[-_\.]?dev[-_\.]?[0-9]*)?         # dev release
                     (?:\+[a-z0-9]+(?:[-_\.][a-z0-9]+)*)? # local
-                    |
-                    \.\*  # Wild card syntax of .*
                 )?
```
AHAHAHAH. We've both spent way-too-long on this.

pradyunsg commented on packaging/_parser.py:
```diff
@@ -0,0 +1,199 @@
+            if not tokens.match("RBRACKET"):
+                tokens.expect("COMMA", error_message="Missing comma after extra")
+                tokens.read("COMMA")
+            if not tokens.match("COMMA") and tokens.match("RBRACKET"):
```
```suggestion
            if tokens.match("RBRACKET"):
```

We don't want to allow `,,,` in extras? This should get a corresponding test too, IMO.

pradyunsg replied:
Ah, nvm. I misread this. :)

pradyunsg reviewed (changes requested):
A few more suggestions. 😅

pradyunsg commented on packaging/_tokenizer.py:
```diff
@@ -0,0 +1,150 @@
+            return self.read()
+        return None
+
+    def raise_syntax_error(self, *, message: str) -> NoReturn:
+        """Raise SyntaxError at the given position in the marker"""
+        at = f"at position {self.position}:"
+        marker = " " * self.position + "^"
+        raise ParseExceptionError(
+            f"{message}\n{at}\n    {self.source}\n    {marker}",
+            self.position,
+        )
```
I really like this! :)

hrnciar commented on packaging/specifiers.py:
```diff
@@ -396,7 +398,10 @@ class Specifier(_IndividualSpecifier):
         """
 
-    _regex = re.compile(r"^\s*" + _regex_str + r"\s*$", re.VERBOSE | re.IGNORECASE)
+    _regex = re.compile(
+        r"^\s*" + _operator_regex_str + _version_regex_str + r"\s*$",
+        re.VERBOSE | re.IGNORECASE,
+    )
```
I split the regex into two parts so I could reuse it in the tokenizer for the VERSION token.
~~~~
"VERSION": re.compile(Specifier._version_regex_str, re.VERBOSE | re.IGNORECASE),
~~~~

brettcannon reviewed (changes requested):
One of the few things that Black can't check syntactically, the docstrings unfortunately don't match the style in the rest of the project (both the newline structure and the sentences need to end with proper punctuation).

hrnciar commented:
I've addressed some remarks, but there is still some stuff to do. I am heading to EuroPython tomorrow so I will continue next week when I am back. If anyone would like to meet in Dublin and have a chat IRL feel free to ping me :).

pradyunsg reviewed (approved):
LGTM, assuming @encukou's suggestion is applied!

hrnciar commented:
I've improved the docstrings and fixed a couple of nitpicks. Unless there is something else to improve, I think, we are done.

hrnciar commented:
@brettcannon could you please do a review? Is there something specific I should improve? I'd like to finish this PR. Thank you.

brettcannon commented:
@hrnciar it's in my review queue, but if @pradyunsg wants to merge w/o a review for me then I am totally fine with that.

hroncok commented:
Changelog entry: https://github.com/pypa/packaging/pull/581

ofek commented:
This is great for CLIs too 🙂 https://github.com/pypa/hatch/issues/449#issuecomment-1236125448
